types/datetime.go

```diff
@@ -0,0 +1,292 @@
+	// Value is a timestamp in milliseconds
+	Value int64
+}
+
+func UnsafeDatetime[T int | int64 | float64](v T) Datetime {
+	return Datetime{Value: int64(v)}
+}
+
+func FromStdTime(t time.Time) Datetime {
+	return Datetime{Value: t.UnixMilli()}
+}
+
+// "YYYY-MM-DD" (date only)
+// "YYYY-MM-DDThh:mm:ssZ" (UTC)
+// "YYYY-MM-DDThh:mm:ss.SSSZ" (UTC with millisecond precision)
+// "YYYY-MM-DDThh:mm:ss(+/-)hhmm" (With timezone offset in hours and minutes)
+// "YYYY-MM-DDThh:mm:ss.SSS(+/-)hhmm" (With timezone offset in hours and minutes and millisecond precision)
+func ParseDatetime(s string) (Datetime, error) {
+	var (
+		year, month, day, hour, minute, second, milli int
+		i                                             int
+	)
+
+	length := len(s)
+
+	if length < 10 {
+		return Datetime{}, fmt.Errorf("%w: string too short", ErrDatetime)
+	}
+
+	// parse YYYY-MM-DD
+
+	// YYYY
+	if !(unicode.IsDigit(rune(s[0])) &&
+		unicode.IsDigit(rune(s[1])) &&
+		unicode.IsDigit(rune(s[2])) &&
+		unicode.IsDigit(rune(s[3]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid year", ErrDatetime)
+	}
+
+	year = 1000*int(rune(s[i])-'0') +
+		100*int(rune(s[i+1])-'0') +
+		10*int(rune(s[i+2])-'0') +
+		int(rune(s[i+3])-'0')
+
+	i = 4
+
+	c := rune(s[i])
+	if c != '-' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	// MM
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid month", ErrDatetime)
+	}
+
+	month = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
+
+	c = rune(s[i])
+	if c != '-' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	// DD
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid day", ErrDatetime)
+	}
+
+	day = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
+
+	if i == length { // done.
+		t := time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
+		return Datetime{Value: t.UnixMilli()}, nil
+	}
+
+	if length < 20 { // can't possibly have a well formed time.
+		return Datetime{}, fmt.Errorf("%w: invalid time", ErrDatetime)
+	}
+
+	// expect 'T'
+	c = rune(s[i])
+	if c != 'T' {
+		return Datetime{}, fmt.Errorf("%w: expected 'T'", ErrDatetime)
+	}
+
+	i++
+
+	// hh
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid hour", ErrDatetime)
+	}
+
+	hour = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
+
+	c = rune(s[i])
+	if c != ':' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	// mm
+
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid minute", ErrDatetime)
+	}
+
+	minute = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
```

Comment:
Maybe it would be nice if there was a few private constants
`
const dateOffset = 3
const hourOffset = 6 
```
etc.

The changing of i, then using offsets from i is a bit hard to read.